incubator/hnc/pkg/validators/object.go

```diff
@@ -0,0 +1,119 @@
+// changes to the objects are allowed.
+//
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=secrets,verbs=create;update,versions=v1,name=secrets
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="rbac.authorization.k8s.io",resources=rols,verbs=create;update,versions=v1,name=roles.rbac.authorization.k8s.io
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="rbac.authorization.k8s.io",resources=rolebindings,verbs=create;update,versions=v1,name=rolebindings.rbac.authorization.k8s.io
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="networking.k8s.io",resources=networkpolicies,verbs=create;update,versions=v1,name=networkpolicies.networking.k8s.io
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=resourcequotas,verbs=create;update,versions=v1,name=resourcesquotas
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=limitranges,verbs=create;update,versions=v1,name=limitranges
+// +kubebuilder:webhook:path=/validate-objects,mutating=false,failurePolicy=ignore,groups="",resources=configmaps,verbs=create;update,versions=v1,name=configmaps
+
+type Object struct {
+	Log     logr.Logger
+	Forest  *forest.Forest
+	client  client.Client
+	decoder *admission.Decoder
+}
+
+func (o *Object) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := o.Log.WithValues("nm", req.Name, "nnm", req.Namespace)
+	inst := &unstructured.Unstructured{}
+	if err := o.decoder.Decode(req, inst); err != nil {
+		log.Error(err, "Couldn't decode request")
+		return deny(metav1.StatusReasonBadRequest, err.Error())
+	}
+	log = log.WithValues("object", inst.GetName())
+
+	oldInst := &unstructured.Unstructured{}
+	// req.OldObject is the existing object. DecodeRaw will return an error if it's empty, so we should skip the decoding here.
+	if len(req.OldObject.Raw) > 0 {
+		if err := o.decoder.DecodeRaw(req.OldObject, oldInst); err != nil {
+			log.Error(err, "Couldn't decode RAW request")
+			return deny(metav1.StatusReasonBadRequest, err.Error())
+		}
+	}
+
+	if isHNCServiceAccount(&req.AdmissionRequest.UserInfo) {
+		log.Info("Allowed change by HNC SA")
+		return allow("HNC SA")
+	}
+
+	resp := o.handle(ctx, log, inst, oldInst)
+	log.Info("Handled", "allowed", resp.Allowed, "code", resp.Result.Code, "reason", resp.Result.Reason, "message", resp.Result.Message)
+	return resp
+}
+
+// handle implements the non-webhook-y businesss logic of this validator, allowing it to be more
+// easily unit tested (ie without constructing an admission.Request, setting up user infos, etc).
+func (o *Object) handle(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured, oldInst *unstructured.Unstructured) admission.Response {
+	oldValue, oldOK := getLabel(oldInst, api.LabelInheritedFrom)
```

Comment:
Nit: I wouldn't call this "OK", what it really means is if the label _exists_ or not. So this should be `oldExists` and `newExists`, that'll make the code below much easier to read.

E.g., instead of:

```
if !newOK {
  verb = "remove"
}
```

we get:

```
if !newExists {
  verb = "remove"
}
```

which reads more clearly (at least to me), because "!exists... remove" are more closely related than "!ok... remove".